go-controller/pkg/util/node_annotations.go

```diff
@@ -555,3 +647,177 @@ func ParseNodeHostAddresses(node *kapi.Node) (sets.Set[string], error) {
+
+	networkID, ok := networkIDsMap[netName]
+	if !ok {
+		return InvalidNetworkID, newAnnotationNotSetError("node %q has no %q annotation for network %s", node.Name, ovnNetworkIDs, netName)
+	}
+
+	return strconv.Atoi(networkID)
+}
+
+// updateNetworkIDsAnnotation updates the ovnNetworkIDs annotation in the 'annotations' map
+// with the provided network id in 'networkID'.  If 'networkID' is InvalidNetworkID (-1)
+// it deletes the ovnNetworkIDs annotation from the map.
+func updateNetworkIDsAnnotation(annotations map[string]string, netName string, networkID int) error {
+	var bytes []byte
+
+	// First get the all network ids for all existing networks
+	networkIDsMap, err := parseNetworkIDsAnnotation(annotations, ovnNetworkIDs)
+	if err != nil {
+		if !IsAnnotationNotSetError(err) {
+			return fmt.Errorf("failed to parse node network id annotation %q: %v",
+				annotations, err)
+		}
+		// in the case that the annotation does not exist
+		networkIDsMap = map[string]string{}
+	}
+
+	// add or delete network id of the specified network
+	if networkID == InvalidNetworkID {
+		delete(networkIDsMap, netName)
+	} else {
+		networkIDsMap[netName] = strconv.Itoa(networkID)
+	}
+
+	// if no networks left, just delete the network ids annotation from node annotations.
+	if len(networkIDsMap) == 0 {
+		delete(annotations, ovnNetworkIDs)
+		return nil
+	}
+
+	// Marshal all network ids back to annotations.
+	networkIdsStrMap := make(map[string]string)
+	for n, id := range networkIDsMap {
+		networkIdsStrMap[n] = id
+	}
+	bytes, err = json.Marshal(networkIdsStrMap)
+	if err != nil {
+		return err
+	}
+	annotations[ovnNetworkIDs] = string(bytes)
+	return nil
+}
+
+// UpdateNetworkIDAnnotation updates the ovnNetworkIDs annotation for the network name 'netName' with the network id 'networkID'.
+// If 'networkID' is invalid network ID (-1), then it deletes that network from the network ids annotation.
+func UpdateNetworkIDAnnotation(annotations map[string]string, netName string, networkID int) (map[string]string, error) {
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	err := updateNetworkIDsAnnotation(annotations, netName, networkID)
+	if err != nil {
+		return nil, err
+	}
+	return annotations, nil
+}
+
+// GetNodeNetworkIDsAnnotationNetworkIDs parses the "k8s.ovn.org/network-ids" annotation
+// on a node and returns the map of network name and ids.
+func GetNodeNetworkIDsAnnotationNetworkIDs(node *kapi.Node) (map[string]int, error) {
+	networkIDsStrMap, err := parseNetworkIDsAnnotation(node.Annotations, ovnNetworkIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	networkIDsMap := map[string]int{}
+	for netName, v := range networkIDsStrMap {
+		id, e := strconv.Atoi(v)
+		if e == nil {
+			networkIDsMap[netName] = id
+		}
+	}
+
+	return networkIDsMap, nil
+}
+
+// NodeZoneAnnotationChanged returns true if the ovnNodeZoneName in the corev1.Nodes doesn't match
+func NodeNetworkIDAnnotation(oldNode, newNode *corev1.Node, netName string) bool {
```

Comment:
`NodeNetworkIDAnnotationChanged`? Reminder to update the comment as well.

Reply:
Ack